src/TgSharp.Core/Session.cs

```diff
@@ -87,14 +138,19 @@ public class Session
         public long LastMessageId { get; set; }
         public int SessionExpires { get; set; }
         public TLUser TLUser { get; set; }
-        private Random random;
+        private readonly Random random = new Random();
+
+        [JsonIgnore]
+        public ISessionStore Store;
 
-        private ISessionStore store;
```

Comment:
is this store->Store change really needed? why make it public if you're adding the `[JsonIgnore]` attrib to it anyway?

Reply:
I made it public so if a user wants to make his own `ISessionStore`, he will be able to set the `Store` property if needed, like [here](https://github.com/Laiteux/TgSharp/blob/master/src/TgSharp.Core/Session.cs#L101) for example.

Reply:
but there's already a Session ctor that receives an ISessionStore instance, you could have used that instead

Reply:
You can only deserialize JSON to parameterless constructors

Reply:
I know? you did this:

```
            var session = JsonConvert.DeserializeObject<Session>(File.ReadAllText(sessionPath), jsonSerializerSettings);

            session.Store = this;
            session.SessionUserId = sessionUserId;

            return session;
```

why not this?:

```
            var deserializedSession = JsonConvert.DeserializeObject<Session>(File.ReadAllText(sessionPath), jsonSerializerSettings);
            var newSession = Session(deserializedSession);
            newSession.SessionUserId = sessionUserId;

            return newSession;
```

Reply:
That would be creating 2 instances of the `Session` class, I prefer my current code

Reply:
Wait, actually your code is invalid, an `ISessionStore` must be passed to the `Session` ctor, what are you trying to achieve?

Reply:
mmm sorry I'm confused now; i just don't understand why the binary store didn't need this field to be public, while yours needs it; public mutable fields are evil, I'm sure it can be done differently

Reply:
Looks like it is because `FileSessionStore` ("binary store") is setting it using the ctor. This might be able to be done differently (not so sure), but trust me we still want this property to be public. I use it for my SQL store I talked to you about a few hours ago, for example, it is required.

Reply:
>  is setting it using the ctor

Let's do the same then.


> I use it for my SQL store

Maybe for your SQL store it's enough to make it protected?

Reply:
> Let's do the same then.

That would require creating a second instance of the `Session` class, only for using the ctor. I don't think this is a good practice. What's so wrong with the current code?

> Maybe for your SQL store it's enough to make it protected?

Should be. You decide.

Reply:
> Should be. You decide.

Let's use `protected` then, public fields are the classic way to shoot yourself in the foot. Creating second instances is not a big deal, our machines are very beefy these days (loss of performance better than loss of maintainability, developer time is more expensive than CPU cycles).

Reply:
I just realized that by creating a new `Session` instance, all property values we got via deserialization will be lost.

Reply:
> we got via deserialization will be lost.

why this doesn't happen in the binary store? should we do a new ctor that lets you do `Session(deserializedSession, this)`?

Reply:
Because the binary store uses this:
```cs
return new Session(store)
{
    AuthKey = new AuthKey(authData),
    Id = id,
    Salt = salt,
    Sequence = sequence,
    LastMessageId = lastMessageId,
    TimeOffset = timeOffset,
    SessionExpires = sessionExpires,
    TLUser = TLUser,
    SessionUserId = sessionUserId,
    DataCenter = defaultDataCenter,
};
```

Reply:
haha

Reply:
actually, that code from the binary store would benefit from the new ctor that I proposed: `Session(deserializedSession, this)`, this way, the vomitive property mapping is only centralized in one place instead of on each store type

Reply:
I thought the same, except the vomitive property mapping doesn't use a `Session` instance but this:
```cs
var id = reader.ReadUInt64();
var sequence = reader.ReadInt32();
var salt = reader.ReadUInt64();
var lastMessageId = reader.ReadInt64();
var timeOffset = reader.ReadInt32();
var serverAddress = Serializers.String.Read(reader);
var port = reader.ReadInt32();
```
😒

Reply:
Back to the beginning. 😕

Reply:
no! you showed me the code before:

```
    AuthKey = new AuthKey(authData),
    Id = id,
    Salt = salt,
    Sequence = sequence,
    LastMessageId = lastMessageId,
    TimeOffset = timeOffset,
    SessionExpires = sessionExpires,
    TLUser = TLUser,
    SessionUserId = sessionUserId,
    DataCenter = defaultDataCenter,
```

that code above would then be in the new ctor

Reply:
Yes! But then how do you call the new ctor from `Session.FromBytes`?

Reply:
(used by both the binary and the json store)

Reply:
holy shit, after doing so much F# programming, I'm starting to realise how much I hate OOP

Reply:
LMAO